### PR TITLE
Make `FlyteFile` compatible with `Annotated[..., HashMethod]`

### PR DESCRIPTION
**Pull request by AdrianoKF**
See issue #3424

# TL;DR
This PR makes `FlyteFile`s usable with offloaded hash calculation via `Annotated[FlyteFile, HashMethod(...)]`, so that they can be cached based on file contents, e.g.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
In `FlyteFilePathTransformer.to_literal()`, `Annotated` type aliases are now
handled by unwrapping the first type argument to the alias, before checking if
they are a subclass of `FlyteFile`.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3424

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/types/file/file.py [hunk 1]
```diff
@@ -272,6 +272,10 @@ def to_literal(
         if python_val is None:
             raise TypeTransformerFailedError("None value cannot be converted to a file.")
 
+        # Correctly handle `Annotated[FlyteFile, ...]` by extracting the origin type
+        if typing.get_origin(python_type) is typing.Annotated:
```
Comment (eapolinario) [inline flytekit/types/file/file.py:276]: import `Annotated` from `typing_extensions` so that we can use it in python 3.8.

**Comment by AdrianoKF** [comment]
> import `Annotated` from `typing_extensions` so that we can use it in python 3.8. 

Fixed here and in test code in aa18baa6

File: flytekit/types/file/file.py (hunk 1 above)
Comment (AdrianoKF) [reply flytekit/types/file/file.py]: Fixed here and in test code in aa18baa6

File: flytekit/types/file/file.py [hunk 2]
```diff
@@ -272,6 +273,10 @@ def to_literal(
         if python_val is None:
             raise TypeTransformerFailedError("None value cannot be converted to a file.")
 
+        # Correctly handle `Annotated[FlyteFile, ...]` by extracting the origin type
+        if typing.get_origin(python_type) is Annotated:
```
Comment (pingsutw) [inline flytekit/types/file/file.py:277]: `get_origin` should also import from typing_extensions

Comment (AdrianoKF) [reply flytekit/types/file/file.py]: Pesky 3.8 compatibility, sorry for the additional turnaround - have been using 3.9+ exclusively for some time, so my mental radar doesn't catch these.

Fixed for `get_origin` and `get_args` in bffa9b8a.

File: tests/flytekit/unit/core/test_flyte_file.py [hunk 3]
```diff
@@ -433,3 +435,19 @@ def wf(path: str) -> os.PathLike:
...
+
+    @task
+    def t1(path: str) -> Annotated[FlyteFile, HashMethod(calc_hash)]:
+        return FlyteFile(path)
+
+    @workflow
+    def wf(path: str) -> None:
+        t1(path=path)
+
+    with tempfile.NamedTemporaryFile() as f:
```
Comment (eapolinario) [inline tests/flytekit/unit/core/test_flyte_file.py:452]: `NamedTemporaryFile` objects are weird on [windows](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) (essentially they can't be opened twice). Specifying `delete=False` should fix the issue.

Comment (AdrianoKF) [reply tests/flytekit/unit/core/test_flyte_file.py]: Sorry for the slow response, I was OOO last week.

I noticed the `local_dummy_file` fixture in the test module and switched to using that instead in c74dca15.

**Review by eapolinario — APPROVED** [review]
